### ISSUE TITLE
afterRefreshCheck infinitely called

### ISSUE THREAD

**Issue by W1nstar** [progress/JSDO#265]
We've been trying to set up a SSO driven JSDO project, but we can't past the fact it hangs on the afterRefreshCheck.

It can login no problem, any CRUD operation is working as expected, but after the token is expired out, the JSDO class tries to refresh it and enters an infinite loop.

We've tried to set this up by ourselves using the scarce documentation we could find on PASOE related SSO. We couldn't find any documentation at all on JSDO with SSO.

Any help would be welcome.

This is how we login on the system:

```
const auth = new progress.data.AuthenticationProviderSSO(serviceURI);

const options = {
    serviceURI: serviceURI,
    authenticationModel: progress.data.Session.AUTH_TYPE_SSO,
    authProvider : auth
};

console.log(auth);

if (!auth._loggedIn) {
	auth.login("walter","walter").then((e) => {
		const jsdoSession = new progress.data.JSDOSession(options);

		jsdoSession.addCatalog(	[catalogURI, 'http://192.10.10.29:8810/web/web/catalog/jsdo/lineasInventarios']).then(data => {
			console.log(data)
			jsdo = new progress.data.JSDO({ name: 'paises' })
			readRecords()
		})
		
	})
	.catch((e)=> console.log(e));
} else {
	const jsdoSession = new progress.data.JSDOSession(options);

	jsdoSession.addCatalog(	[catalogURI, 'http://192.10.10.29:8810/web/web/catalog/jsdo/lineas']).then(data => {
		console.log(data)
		jsdo = new progress.data.JSDO({ name: 'paises' })
		readRecords()
	})
}
```

After the token has expired, this is the callstack we're facing:

[the callstack](https://imgur.com/a/PuSelXX)
[this is how it starts](https://imgur.com/a/WA5sNLK)

Like I said, we're doing this by ourselves so any help would be appreciated.

Thanks in advance, and apologies if this is common knowledge but we're just beign dumb.

**Comment by W1nstar** [comment]
I might have found the culprit, but I lack enought comprehension of the whole JSDO to say if it's ok.

**I'd still like to know what do you guys think! I'm not enterely sure of what I'm doing!**

When the AuthenticationProviderSSO is created, there's this piece of code about credentials that is seemingly rewriting a callback, thus creating the infinite loop. If I rename that var, it all seems to work. It refreshes as expected, or at least it seems so. This is the piece of code:

```
            if (this.hasClientCredentials()) {
                // Every token given has an expiration "hint". If the token's lifespan
                // is close to or past that limit, then a refresh is done.
                // No matter what the outcome of the refresh, keep in mind we always 
                // send the original request.
                date = new Date();
                if (this.automaticTokenRefresh &&
                    this.hasRefreshToken() &&    
                    date.getTime() > retrieveAccessTokenExpiration()) {
                    try {
                        var callback_mks = function (params, result, info) { //RENAMING THIS FROM callback to callback_mks SORRY
                            params = progress.util.Deferred.getParamObject(params, result, info);
                            afterRefreshCheck(params.provider, params.result, params.info);
                        };
                        // finally
                        this.refresh()
                            .then(callback_mks, callback_mks); //CALLING MY RENAMED VAR DON'T MIND ME
                    } catch (e) {
                        callback_mks(e); //MORE RENAMING
                    }
                } else {
                    afterRefreshCheck(this, progress.data.Session.SUCCESS, null);
                }
            } else {
                // This message is SSO specific, unless we can come up with a more general message 
                // JSDOSession: The AuthenticationProvider needs to be managing a valid token.
                errorObject = new Error(progress.data._getMsgText("jsdoMSG125", "AuthenticationProvider"));
                callback(errorObject);
            }
```

I hope someone could help me.

**Comment by joshualan** [comment]
Hey, @W1nstar, the JSDO unfortunately does not officially support SSO.... though if we have enough interest, i can definitely bring this up as something we can officially support.

I think that we might've accidentally brought us into callback hell where the name `callback` got overloaded, causing it to go into a loop? 

You might just write `jsdosession.automaticTokenRefresh = false;` to temporarily get aroundthis and manually call `refresh()` every now and then for now. 

I'll try to take a look tomorrow

**Comment by W1nstar** [comment]
Thanks a lot for the response. You really made my day! I look forward your help.

The callback hell might be real. With the change I did works, but sometimes errors somewhere else, so I guess I'm not on the right track there. I tried to personally look into this, and my goal was to make you guys a PR, but as I said I lack knowledge on most themes going on here to help you. I'll need a few days to be able to look at it, but I will try. We're trying to avoid a manual refresh.

We were under the impression it was supported since it seemed to have everything needed inside, plus we saw [advertising](https://d117h1jjiq768j.cloudfront.net/docs/default-source/default-document-library/introduction-to-telerik-platform-for-mobile-development.pdf?sfvrsn=d4c8000c_0) on progress's learning courses about pasoe, jsdo and SSO. It's a bummer it's not supported officially. Is there any way we can express our interest more _officially_?

Thank you again for your feedback, @joshualan .

**Comment by W1nstar** [comment]
Any news on this issue, @joshualan ? I was able to stop the infinite loop, but now we have found there's an unhandled exception (https://github.com/progress/JSDO/blob/e4461620ecfcc71b2e6e95dd89c22aa38e083bee/lib/progress.all.js#L10207 it even has "we have to fix this" comment) when the token is rejected, that is proving very difficult for us to fix it without full knowledge of what's going on. I've already spent a full day trying things, but nothing seems to work.
Edit: forgot to say, the unhandled exception is not a part of sso, it's on the general _openRequest.

**Comment by joshualan** [comment]
@W1nstar we're throwing something in the backlog for this so we might have time to look into this in the next few months but we judge the amount of interest in the project via pull requests.

Sorry for the late reply, was on vacation for a while. 

Is the same original code causing this issue?

**Comment by W1nstar** [comment]
@joshualan yeah, on the original code there's still the infinite loop plus the _openRequest throw arg.

**Comment by seankupiec** [comment]
This fix was the same as mine as well. The deceleration of "callback" inside _openRequestAndAuthorize function caused a loop from afterRefreshCheck at line 291. Updating line 308 and 314 corrects this issue and successfully refreshes the token and continues the original request in our project.  

@edselg  Do you think we can get the pull request #276 pushed into the live branch for others to be able to use it (including our other projects)?
